README
$Header: /CVSROOT/scylla-charybdis/README,v 1.7 2003/11/02 22:54:05 tino Exp $

Name:
	Scylla and Charybdis (with Odysseus)

Version:
	2.3.0 (interim with several new features and scripts)
	Scylla and Charybdis is developed with the "Release early,
	release often" paradigma in mind.  However I have very few time,
	so "often" probably means 1-2 times in a century, ok?

License:
	This software comes with absolutely no warranty of any kind.
	It may be distributed under the terms of the
	GNU General Public License (GPL) v2 or higher.
	See file COPYING for details.

Purpose:
	To get a data odyssey done with the help of two small monsters.
	Backup/Transfer files from one host to another.
	Limited restore capability with little_poseidon.sh helper script.
	Read the new file RATIONALE for why I programmed S&C.

Instable:
	Version 2.3.0 has some new features and is considered
	nonstable.  Changes for large file support renders elder
	scylla/charybdis incompatible on files of 1 GB size or bigger.

Stable:
	Version 2.2.5 is stable as it is - without change - based on
	Version 2.2.4 which has transferred gigabytes successfully.

Untested:
	examples/little_poseidon.awk
	Script to re-create the filesystem metadata from odysseus output.

Features:
	Scylla must be run from inetd.
	Charybdis connects to Scylla to upload files.
	The protocol uses ZLIB (compression) and SSL (encryption).
	If a partial uploaded file is found the missing part is appended.
	If a correctly uploaded file is found charybdis gives exit code 0.
	The upload is protected by a clear text (SSL transmitted) password.
	Heavy timeout protection is built in in case that dialup breaks.
	Charybdis shows a little upload progress meter on stdout.
	There is a throtteling feature such that Charybdis does not
	utilize the full bandwidth.

Changes:
	see file HISTORY

Notes:
	Especially useful with auto-dialing connections.
	Upload-success is detected on file size and MD5-HASH.
	Throtteling is not accurate.  Look at the progress meter!
	Currently hard coded timeout values are built in.
	You have to connect at least twice to Scylla to find out the
	success of the upload, so perhaps use following loop:

	# example corrected for v2.1
	for file in *
	do
		while charybdis host port password "$file"
		do
			:
		done
	done

	LINEFEED in filenames are not yet supported.
	This is an absolute BETA!
	All documentation is the source!

Install:
	You need to have installed following packages:
		openssl-devel libz-devel gdbm
	Then type:
		make
	See example/make-ssl-certificate on how to create /etc/scylla.pem
	If anything does not work you are on your own
	(-DNO_SSL and -DNO_ZLIB probably don't work anymore).

Cygwin:
	If it does not compile under your Cygwin, have a look in
	sc-cygwin.h
	Cygwin needs some real weird treatment, not because of the
	port, only because of the includes that come with Cygwin.

Security:
	Scylla shall run as nonprivileged user, like nobody.
	Charybdis must have access to all files.
	Odysseus must be able to read all directories.

Usage:
	scylla directory password [sslcertificate]
		Run Scylla (this here) from inetd, see inetd.conf below.
		It accepts an encrypted and compressed data stream.
		Scylla assumes that Charybdis feeds a file.
		Scylla writes the file into the given directory.
		Missing sub-directories are created as needed.
		Existing (changed) files are renamed.
		Return values:
			meaningless

	charybdis host port password filename [throttle [blocksize]]
		Charybdis (this here) assumes that scylla is at host:port
		It opens an compressed and encrypted data stream to Scylla.
		Charybdis authenticates at Scylla with the password.
		Then the file is uploaded to Scylla OR (else)
		it is detected that the file is corrctly at the other side.
		Run this in a shell loop.
		Broken uploads are restarted seamlessly.
		The bandwidth can be controlled with trottle@blocksize
		Throttle is given in KB/s
		Return values:
			0	file is already known & unchanged
			1	unreadable files
			2	file was uploaded
			3	files uploaded and unreadable files
			else	error/timeout/etc.

	charybdis host port password - [throttle [blocksize]]
		As before, but files are streamed.
		This highly improves throughput as this skips
		quite some SSL overhead.
		However it currently does not work like tar | gzip

	odysseus directory...
		Odysseus ist pre alpha.  It traverses the directories on the
		commandline and prints out all file and non-file stats it finds.
		Directory must be the path which is printed by /bin/pwd.
		If a directory differs from getcwd() odysseus bails out for
		security reasons (i. E. directory is moved while accessed).

Example:
	Read README.backup and README.restore.  Have a look into example/

	/etc/inetd.conf:
	999 stream tcp nowait.999 scylla /usr/sbin/tcpd /usr/local/sbin/scylla /scylla change-this-password /etc/scylla.pem

	V2.3.0 example for backup:

	su -
	mkdir backup &&
	/path/to/source/of/scylla-charybdis/example/backuploop.sh back.up.host:999 /

	V2.2.2 example for backup with file streaming mode (one liner with auto retry):

	lst="{home,var,boot,,usr,opt,}";
	odysseus / > `hostname`.odysseus; b="backup.`hostname`";
	ln -s / "$b"; while eval ! find \"\$b\"/$lst -type f -xdev -print |
	/usr/local/sbin/charybdis host.domain.tld 999 change-this-password -;
	do sleep 500; done

	V2.1 and before example:
	# !!!WARNING!!!
	# Following kills a file after a successful upload:
	# !!!WARNING!!!
	# You have to run this twice.
	find * -type f -print |
	while read -r a
	do
		while [ -f "$a" ]
		do
			/usr/local/sbin/charybdis host.domain.tld 999 change-this-password "$a" &&
			rm -vf "$a"
		done
	done

Caveats:
	Scylla refuses to accept files which start with / or have .. as a
	path component, thus you cannot give absolute files to Charybdis.

	Scylla writes a logfile to /var/tmp/scylla.log.  This log can
	become quite big.  To protect against this make it
	unwriteable like with "mkdir /var/log/scylla.log"

	Charybdis reads the files, thus alters the access time of the files.

	example/little_poseidon.sh *is*not*tested*, sorry.

Todos:
	Finish examples/little_poseidon.sh and
	README.backup on how to backup a Linux completely,
	README.restore on how to restore it from scratch

	V2.3.x will be the last release of Scylla+Charybdis.

Unlikely:
	Use chroot() in scylla.
	Do not restart compression on each file.
	Better scylla logging.
	Transport file flags (like tar).
	SSL authentication and reverse (download) operation.
	Public Upload/Download (skip password) Option.
	MTU hacks and UDP option (backup succeeds on really weird links).

	DBM support which will lead to:
	Hephaestos (compressed file storage)
	Options for Odysseus like -xdev or odysseus | charybdis

Bugs:
	LINEFEED in filenames are not yet supported!
	Streaming mode garbles speedometer output of charybdis.
	Streaming mode currently resets zlib compression for each file.
	Throtteling is not very accurate.
	Charybdis tells some crap like 'protocol failure' instead of telling
	about SSL problems if something fails (unreadable /etc/scylla.pem).
	S&C currently has no sparse file support.
	In stream mode there is no real grace if something breaks.
	Only files are supported, hardlinks are copied individually,
	softlinks are copied as the file the softlink points to.
	Version 2.2 really does not work well with V2.1.
	Version 2.3 is incompatible to V2.2 on files bigger than 2GB.

Observations:
	I observed once scylla+charybdis did disagree on a MD5 sum on an
	identical file, such that the data already present was renamed
	and uploaded again.  I canceled the transfer, renamed the file
	back an started it again.  Now both detected that the file needed
	no upload again.  Maybee it's due to the fact that the *receiving*
	machine has NON ECC RAM such that this was a bit glitch in main
	memory.  Anyway, no harm was done, only something was transferred
	again which was already there correctly.  Note that I never observed
	the opposit, so charybdis never told, a file was successfully
	uploaded while this was not the case.  However I always transfer
	more data than the cache size of the machine, such that scylla must
	re-read from disk when MD5 sums are calculated.  However this
	thoughts leave room for improvement for the paranoid (like me):
	Scylla should write out files synced with RAM caching disabled.
	But how to obtain this in a portable way?
	As soon as files are stored under the MD5 filename this will become
	less 'urgent', though.

History:
 	In V2.2 there was a serious memory leackage, fixed in V2.2.1.
	V2.2.4 now compiles and was tested under RedHat 8.0, too.
	V2.2.5 now compiles and was tested under RedHat 9.0, too.
	V2.3.0 got big file and starts to need GDBM library now.
	Read README.backup and README.restore on how to do a restore.

Roadmap:
	No roadmap any more, sorry.
	V2.3.x will be the very last release before a complete
	rewrite, which will be released under a new name.
	The homepage of S+C will lead you to the successor.
	However, this rewrite may take years, as always.

Source:
	Homepage: http://www.scylla-charybdis.com/
	Latest:   http://www.scylla-charybdis.com/scylla-charybdis.tgz

Copyright:
	Copyright (C)2000-2003 by Valentin Hilbig
	diskimg-reply@03.softkill.org
	(My real eMail address is not listed on public media due to SPAM.)
	http://geht.net/tino/tino.html
	(I am natively German speaking)

$Log: README,v $
Revision 1.7  2003/11/02 22:54:05  tino
Cygwin special treatment

Revision 1.6  2003/11/02 22:38:46  tino
Updated for the upcomming release



HISTORY
$Header: /CVSROOT/scylla-charybdis/HISTORY,v 1.5 2003/11/02 22:38:11 tino Exp $

2003-10-21 2.3.0:
	Version 2.3.x will be the last version track.  Perhaps there
	will be a successor sometimes with a new name.  Note that I am
	planning a public CVS where you can access even non-released
	versions.  As I do not know any way to sync two CVS
	repositories automagically, I cannot use CVS at sourceforge yet.

	New script:
	example/backuploophelper.sh

	New nonstable features:
	Scylla now also knows SSL-less mode (empty PEM-File argument).
	MD5-directory splitmode added when DIR starts with MD5/ (where
	MD5 is ignored silently).  Large files are stored under their
	MD5 sum and splited into 1 GB chunks - this needs DBM.
	example/little_poseidon.sh does not work with MD5 directories,
	so on restores you are at your own currently.

	Local file append bug (possible exploit) removed:
	The very basic logging has been moved from /tmp/scylla.log to
	/var/log/scylla.log

2003-10-02 2.2.5 (never released publicly):
	Two helper scripts added.  One as a workaround for the missing
	poseidon:
	example/backuploop.sh (which runs without problems)
	example/little_poseidon.sh (which is nearly untested)
	See the conceptual papers README.backup and README.restore on
	how to use this scripts to do a backup and restore.
	EXCEPT FOR BUGFIX RELEASES THERE WILL BE NO FURTHER
	DEVELOPEMENT OF VERSION 2.x OF S+C!

2003-05-06 2.2.4 (not announced on freshmeat):
	This version is declared stable.  Draft of some concept papers
	added.
	However it looks like this is the last version of this type of
	S&C as I am in the planning phase of version 3.0, which will
	be a complete rewrite with a major structure change.
	So do not expect Poseidon/restore capability, sorry.

2003-03-18 2.2.3:
	A little bug in Charybdis removed which broke stream mode
	on missing files.
	Put version on website.

2002-12-05 2.2.3:
	Improved Odysseus output (device)
	Improved Charybdis progress meter.
	Improved counting of throughput.
	Not available as download, as my web server crashed.
	S+C of this version still cannot handle files >=2GB.

2002-11-05 2.2.2:
	First version of Odysseus, started source redesign.
	Was available in download, but not tested and not announced.
	It was then tested and stability (odysseus, too) proved ok.

2002-07-14 2.2.1:
	Memory leakage resolved.  Esp. Charybdis ate a lot of memory.
	File streaming mode should be a lot more usable now ;)
	Wohoo, there is now a homepage!

2002-06-18 2.2:
	Charybdis reads a file list from stdin if filename = -
	This feature is highly experimental. (Fixed for 2.2.1)
	BUGfix in newname(): would never reach .~03~

2002-05-09 2.1:
	Charybdis now sets MTU.
	Scylla now refers to /etc/scylla.pem by default
	Scylla writes some diagnostics to /tmp/scylla.log
	Domain registered and documentation updated.

2001-06-18 2.0: (actually there are two versions, I count the later one)
	Scylla now needs OpenSSL and ZLIB installed.
	Charybdis does not eat (removes) files any more.
	It returns 0 if everything has been transferred successfully
	and returns 2 if the file has been transferred (success unknown).
	The throtteling code has been revised, too.

2000 1.0: Initial version without ZLIB and SSL.
	Charybdis eats away (removes) files after successful upload.
	This version was never released publicly.

$Log: HISTORY,v $
Revision 1.5  2003/11/02 22:38:11  tino
changed to describe 2.3.0

Revision 1.4  2003/10/22 03:26:04  tino
Taken first steps for LONGFILE support and MD5-Directory mode



charybdis.c
/* $Header: /CVSROOT/scylla-charybdis/charybdis.c,v 1.3 2003/11/06 06:57:49 tino Exp $
 *
 * $Log: charybdis.c,v $
 * Revision 1.3  2003/11/06 06:57:49  tino
 * Customizeable timeout added (300 was not enough for me now ..)
 *
 * Revision 1.2  2003/11/02 22:21:37  tino
 * CVS keywords added
 *
 */
#define	HERE_CHARYBDIS

#define	CHARYBDIS_TIMEOUT	999	/* Increased timeout	*/
int	charybdis_timeout;

#include "util.h"

#include <netinet/tcp.h>
#include <time.h>

static unsigned long	totalcount, overflowcount;
static char		workmode;
static int		returnval;
static int		filecount, filecopy;

#define	INTERPOLATION 30

static void
alarm_ticker(void)
{
  static int	speedometer[INTERPOLATION], speed2[INTERPOLATION];
  static time_t	timing[INTERPOLATION];
  static unsigned long	lastcount, lastio;
  static unsigned long	gigabytes;
  static long	delta, delta2;
  static int	pos;
  time_t	now;
  long		seconds;
  double	bytes, bytes2, percent;

  pos			= (pos+1)%INTERPOLATION;

  time(&now);

  if (overflowcount)
    {
      lastcount		= totalcount-overflowcount;
      overflowcount	= 0;
    }
  if (copycount_out>=(1ul<<30))
    {
      gigabytes++;
      copycount_out	-= 1ul<<30;
      lastio		-= 1ul<<30;
    }
  delta2		-= speed2[pos];
  delta2		+=
    speed2[pos]		= copycount_out-lastio;
  lastio		= copycount_out;

  delta			-= speedometer[pos];
  delta			+= 
    speedometer[pos]	= copycount-lastcount;
  lastcount		= copycount;

  seconds		= now-(timing[pos] ? timing[pos] : timing[1]);
  timing[pos]		= now;
  if (seconds<=0)
    seconds		= 1;

  bytes			= (double)delta/seconds/1024.;
  bytes2		= (double)delta2/seconds/1024.;
  percent		= -1;
  if (totalcount)
    percent		= copycount*100./totalcount;

  printf("\r%c %7ld %6.2f%% %8.2f %8.2f %2d %d %d %lu.%05lx",
	 workmode, (totalcount-copycount)>>10,
	 percent, bytes, bytes2, alarm_timeout,
	 filecount, filecopy,
	 gigabytes, copycount_out>>10);
  fflush(stdout);
}

static void
getaddr(struct sockaddr_in *sin, const char *host, int port)
{
  memset(sin, 0, sizeof *sin);
  sin->sin_family	= AF_INET;
  sin->sin_addr.s_addr	= htonl(INADDR_ANY);
  sin->sin_port		= htons(port);
  if (host && !inet_aton(host, &sin->sin_addr))
    {
      struct hostent	*he;

      if ((he=gethostbyname(host))==0)
	ex(host);
      if (he->h_addrtype!=AF_INET || he->h_length!=sizeof sin->sin_addr)
	ex("unsupported host address");
      memcpy(&sin->sin_addr, he->h_addr, sizeof sin->sin_addr);
    }
}

static void
sockopen(const char *host, int port)
{
  struct sockaddr_in	sin;
  int			on;
  struct linger		linger;
  unsigned int		mss;

  sock	= socket( AF_INET, SOCK_STREAM, 0 );
  if (sock<0)
    ex("socket");

  on = 1;
  if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &on, sizeof on))
    ex("setsockopt reuse");
  linger.l_onoff	= 1;
  linger.l_linger	= 65535;
  if (setsockopt(sock, SOL_SOCKET, SO_LINGER, &linger, sizeof linger))
    ex("setsockopt linger");
  on	= 102400;
  if (setsockopt(sock, SOL_SOCKET, SO_SNDBUF, &on, sizeof on))
    ex("setsockopt sndbuf");
  on	= 102400;
  if (setsockopt(sock, SOL_SOCKET, SO_RCVBUF, &on, sizeof on))
    ex("setsockopt rcvbuf");

  /* 2002-05-09 set the MTU
   * The kernel takes a lower value if it is not supported ;)
   * We add some bytes for the tcp packet overhead.
   */
  if (maxblock<1500)
    {
      mss = maxblock+48;
      if(setsockopt(sock, IPPROTO_TCP, TCP_MAXSEG, &mss, sizeof(mss)))
        ex("setsockopt maxseg");
    }

  getaddr(&sin, host, port);

  if (bind(sock, (struct sockaddr *)&sin, sizeof sin))
    ex("bind");

#if 0
  if (fcntl(sock, F_SETFL, O_NDELAY))
    ex("O_NDELAY");
#endif
}

static void
do_connect(const char *host, int port)
{
  struct sockaddr_in	sin;

  sockopen(NULL, 0);

  getaddr(&sin, host, port);
  if (connect(sock, (struct sockaddr *)&sin, sizeof sin))
    ex("connect");
}

/* We don't want to connect to scylla
 * except we have to do something.
 * Therefor this delayed connect function.
 * Additionally this prohibits a timeout
 * until find feeds us the files ;)
 */
void
delayed_connect(const char *host, int port, const char *pass)
{
  static char	*sav_host, *sav_pass;
  static int	sav_port;

  DP(("delayed_connect(%s,%d,%s)", host, port, pass));
  if (host)
    {
      sav_host	= str_alloc(host);
      sav_port	= port;
      sav_pass	= str_alloc(pass);
      sock	= -1;
      return;
    }
  if (sock>=0)
    return;

  workmode	= '0';
  alarm_init(1 /* see below */, charybdis_timeout);

  /* Hack: Set a connect timeout to 5 minutes
   */
  alarm(300);
  do_connect(sav_host, sav_port);

#ifndef NO_SSL
  alarm_process();
  alarm(300);
  tino_ssl_client();
#endif

  /* Reset timeout to alarm processing
   */
  alarm_reset();
  alarm_process();
  alarm(1 /* see above */);

  workmode	= '1';

  /* send password */
  swrite(sav_pass);

  if (sread_match("scylla", 6))
    ex("where is my dear scylla?");
  /* leave version control for future
   * we are using version 1.0 protocol
   * (plain password)
   */
}

void
dofile(const char *file)
{
  int		fd;
  char		*t;
  unsigned long	len;
  struct stat	st;

  DP(("dofile(%s)", file));

  filecount++;
  alarm_reset();
  alarm_process();

  if ((fd=open(file, O_RDONLY))<0)
    {
      perror(file);
      DP(("dofile() !file"));
      returnval	|= 1;
      return;
    }
  if (fstat(fd, &st) ||
      !S_ISREG(st.st_mode))
    {
      perror(file);
      DP(("dofile() !fstat"));
      close(fd);
      returnval	|= 1;
      return;
    }

  /* Say there is more on streaming mode
   * On the first file delayed_connect is done,
   * thus sock<0 here
   */
  if (sock>=0)
    swrite("m");

  /* Now that we have something to do: connect
   */
  delayed_connect(NULL, 0, NULL);

  workmode	= '2';
  swrite(file);

  t	= sread();
  if (!isdigit(*t))
    {
      len	= 0;
      swrite("c");
      workmode	= 'c';
    }
  else
    {
      char	*m;

      sscanf(t, "%lu", &len);
      m	= md5sum(fd, len);
      if (!m)
	ex("sum");

      if (!strcmp(m, t))
	{
	  if (st.st_size==len)
	    {
	      free(m);
	      free(t);
	      close(fd);
	      swrite("o");
	      /* unlink(argv[4]); */
	      return;
	    }
	  swrite("a");
	  workmode	= 'a';
	}
      else
	{
	  len	= 0;
	  swrite("w");
	  workmode	= 'w';
	}
      free(m);
    }
  free(t);

  filecopy++;

  if (lseek(fd, len, SEEK_SET)!=len)
    ex("lseek");
  overflowcount	+= copycount;
  copycount	= 0;
  totalcount	= st.st_size-len;

  if (sread_match("OK", 0))
    ex("protocol failure");

  if (docopy(fd, sock, 0))
    ex("copy error");
  if (close(fd))
    ex("close error");

  returnval	|= 2;
  workmode	= 'e';
  /* Wait for transfer flush */

  if (sread_match("NEXT", 0))
    {
      fprintf(stderr, "protocol mismatch: missing next\n");
      exit(returnval);
    }
}

/* TODO in 2.3:
 * We are not allowed to buffer, as if something breaks,
 * we must leave the other files on stdin.
 * Therefor we must loop over bytes.
 * Fix this with S&C version 3 (or 4?).
 */
void
stdinfileloop(void)
{
  char	buf[BUFSIZ];

  000;
  DP(("stdinfileloop"));
  while (fgets(buf, sizeof buf-1, stdin))
    {
      size_t	len;

      len	= strlen(buf);
      if (len>=sizeof buf-1)
	ex("filename too long");
      if (len && buf[len-1]=='\n')
	buf[--len]	= 0; 
      if (len<1)
	continue;

      dofile(buf);
    }
}

int
main(int argc, char **argv)
{
  DP(("main(%d,%p)", argc, argv));
  if (argc<5)
    {
      printf("Usage: %s host port password filename [throttle [blocksize [timeout]]]\n"
	     "\tCharybdis (this here) assumes that Scylla is at host:port\n"
	     "\tCharybdis authenticates at Scylla with the password.\n"
	     "\tWhen the file is uploaded to Scylla it returns 2 or\n"
	     "\tif Scylla already has the file completely it returns 0.\n"
	     "\tSo run this in a shell loop until true is returned\n"
	     "\tto restart broken uploads seamlessly.\n"
	     "\tThe bandwidth can be controlled with trottle@blocksize\n"
	     "\t\t(throttle is in KB/s)\n",
	     argv[0]);
      return 1;
    }

  returnval	= 0;
  if (argc>5)
    {
      throttle	= atoi(argv[5]);
      maxblock	= 512;
      if (argc>6)
	maxblock	= atoi(argv[6]);
    }
  if (maxblock<100)
    maxblock	= BUFSIZ;
  else if (maxblock<256)
    maxblock	= 256;

  charybdis_timeout	= CHARYBDIS_TIMEOUT;
  if (argc>7)
    charybdis_timeout	= atoi(argv[7]);
  if (charybdis_timeout<=0)
    charybdis_timeout	= CHARYBDIS_TIMEOUT;

  delayed_connect(argv[1], atoi(argv[2]), argv[3]);

  if (strcmp(argv[4], "-"))
    dofile(argv[4]);
  else
    stdinfileloop();

  if (sock>=0)
    {
      swrite("e");

#ifndef NO_SSL
      tino_ssl_close();
#endif
      close(sock);
    }
  return returnval;
}
